pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -74,16 +76,98 @@ type cloudWatchExecutor struct {
 
 func (e *cloudWatchExecutor) newSession(region string) (*session.Session, error) {
 	dsInfo := e.getDSInfo(region)
-	creds, err := getCredentials(dsInfo)
+
+	cacheKey := fmt.Sprintf("%d:%s:%s:%s:%s:%s", dsInfo.AuthType, dsInfo.AccessKey, dsInfo.Profile, dsInfo.AssumeRoleARN,
+		dsInfo.ExternalID, region)
+
+	sessCacheLock.RLock()
+	if env, ok := sessCache[cacheKey]; ok {
+		if env.expiration.After(time.Now().UTC()) {
+			sessCacheLock.RUnlock()
+			plog.Debug("Using cached AWS session", "key", cacheKey)
+			return env.session, nil
+		}
+	}
+	sessCacheLock.RUnlock()
+
+	cfgs := []*aws.Config{
+		{
+			CredentialsChainVerboseErrors: aws.Bool(true),
+		},
+	}
+
+	var regionCfg *aws.Config
+	if dsInfo.Region == defaultRegion {
+		plog.Warn("Region is set to \"default\", which is unsupported")
+		dsInfo.Region = ""
+	}
+	if dsInfo.Region != "" {
+		regionCfg = &aws.Config{Region: aws.String(dsInfo.Region)}
+		cfgs = append(cfgs, regionCfg)
+	}
+
+	switch dsInfo.AuthType {
+	case authTypeSharedCreds:
+		plog.Debug("Authenticating towards AWS with shared credentials", "profile", dsInfo.Profile,
+			"region", dsInfo.Region)
+		cfgs = append(cfgs, &aws.Config{
+			Credentials: credentials.NewSharedCredentials("", dsInfo.Profile),
+		})
+	case authTypeKeys:
+		plog.Debug("Authenticating towards AWS with an access key pair", "region", dsInfo.Region)
+		cfgs = append(cfgs, &aws.Config{
+			Credentials: credentials.NewStaticCredentials(dsInfo.AccessKey, dsInfo.SecretKey, ""),
+		})
+	case authTypeDefault:
+		plog.Debug("Authenticating towards AWS with default SDK method", "region", dsInfo.Region)
+	default:
+		panic(fmt.Sprintf("Unrecognized authType: %d", dsInfo.AuthType))
```

Comment:
@marefr I don't agree that whether to use `panic` or not should depend on whether the enclosing function returns an error. If a function doesn't return an error, and you need to return one you should add `error` to the return signature.

`panic` is for special cases, where it makes sense to crash the program and get a stacktrace. You can read about how `panic` should be used [here](https://golang.org/doc/effective_go.html#panic). The question is whether it makes sense to panic here. I _think_ it does, since this branch will only execute in case of a hard error made by the programmers (us) and I think it's helpful to not try to continue, but dump the stack so we can easily diagnose what led to it.

Reply:
It's more if we know this would happen it would make sense to return an error instead of getting a full stacktrace and let our recovery middleware handle it since it will not crash the Grafana process. Maybe we know this should never happen and in that case please disregard this comment.

Reply:
@marefr That's exactly the case here, it's a branch of code that should never execute. If it does, an auth type has been added that this switch statement doesn't handle and we should fix it ASAP.